### ISSUE TITLE
Semantic Segmentation & DALIGenericIterator

### ISSUE THREAD

**Issue by jasonxzhou** [NVIDIA/DALI#3832]
Hi DALI team,
Thanks for your work on DALI - I had a few questions about our particular use-case of segmentation. For us, each sample consists of 2 images (RGB image, and ground-truth pixel-wise annotation image), and we train with `ddp` in PyTorch-Lightning (using sharding, etc.).

Because `DALIGenericIterator` only takes in a single `reader_name`, I need to define 1 pipeline for the RGB image, 1 pipeline for the pixel-wise annotation images, and pass both in a list as the `pipelines` param to `DALIGenericIterator`. Consequently, I cannot sync particular random transforms across the two pipelines (for example, randomly horizontally flipping a particular sample would involve flipping both the RGB and the annotation image). 

Is there a way to do this as part of DALI in particular? I would be able to get around this if `DALIGenericIterator` accepted multiple `reader_names` and could shard them identically, or perhaps I'm missing something about synchronizing randomness across two pipelines? (perhaps I can use 1 pipeline's randomness as input to the other pipeline to sync them?)

(I could also apply those transforms after retrieving them from DALI using standard PyTorch transforms, but would like to keep all augmentations in same place with DALI)

**Comment by JanuszL** [comment]
Hi @jasonxzhou,

The `reader_name` parameter in the iterator is used to infer the size of the epoch and drive the framework DALI iterator. However, you can have more than one reader in the pipeline itself. To make sure they are in sync please provide the same number of samples and make sure that the seed is also the same. Something like this should work:
```
@pipeline_def
def simple_pipeline():
    jpegs_images, _= fn.readers.file(files=images_files, seed=1234, name="main_reader")
    jpegs_masks, _= fn.readers.file(files=masks_files, seed=1234)
    images= fn.decoders.image(jpegs_images, device='mixed')
    masks= fn.decoders.image(jpegs_masks, device='mixed')
    return images, masks

pipe = simple_pipeline(batch_size=max_batch_size, num_threads=1, device_id=0)
iter = DALIGenericIterator([pipe], ['data', 'label'], reader_name='main_reader')
```

**Comment by jasonxzhou** [comment]
Thanks @JanuszL for the quick response. Would this handle sharding + ddp properly? (sharding the image and masks identically)

**Comment by JanuszL** [comment]
Adding `num_shards` and `shard_id` arguments should do:
```
@pipeline_def
def simple_pipeline():
    jpegs_images, _= fn.readers.file(files=images_files, seed=1234, name="main_reader", num_shards=world_size, shard_id=global_rank)
    jpegs_masks, _= fn.readers.file(files=masks_files, seed=1234, num_shards=world_size, shard_id=global_rank
    images= fn.decoders.image(jpegs_images, device='mixed')
    masks= fn.decoders.image(jpegs_masks, device='mixed')
    return images, masks

pipe = simple_pipeline(batch_size=max_batch_size, num_threads=1, device_id=0)
iter = DALIGenericIterator([pipe], ['data', 'label'], reader_name='main_reader')
```

**Comment by jasonxzhou** [comment]
Thanks very much!
